Add an API for plugins

Akul2010 opened the pull request:
The previous PR, #12, was discontinued because I messed up somewhere and wasn't able to do anything about it. But, I have ported the same changes to this new version with less messy commits.

This basically gives all plugins access to the speak and listen functions when run, so they too can speak and listen when needed rather than printing output in CMD or listening once without follow-up questions.

Akul2010 committed:
Get the API set up and remove unneeded `start` function

Akul2010 committed:
Update paths in API for simplicity

Akul2010 committed:
Add fastapi to the dependencies

Akul2010 committed:
Delete makefile as it was unneeded

Akul2010 committed:
Update create_plugins.md to include code for API

Akul2010 commented:
Hi @aaronchantrill, I have made some changes and would like you to review them as I am not as experienced with Perl/JS as I am with Python, so I'm not sure if I got anything wrong. Please try to doit as soon as possible so that I can update the plugins in the akulai-plugins repository when finished.

Akul2010 committed:
Use sys.exit() rather than exit()

aaronchantrill reviewed (changes requested):
I think I understand what you are doing here, but I'm getting some errors.

I had to move the self... lines just after `if __name__ == '__main__'` into the init method of the AkulAI() class. Now I'm getting "Error loading ASGI app. Attribute "app" not found in module "akulai". I have tried both running `akulai.py` and running `uvicorn akulai:app --reload` and got the same message either way. Do you have some special way you are running this?

I'm not familiar with FastAPI, so it will take me a little time to figure out, so please let me know if this makes sense to you.

Thanks!

aaronchantrill commented:
Okay, I think I understand what the problem was. I had to move the api setup stuff above the if __name__ == __main__ but below the AkulAI() class definition. Then I changed it to use different paths for speak and listen:
```
@app.get("/speak/{text}")
async def speak(text: str):
    akulai.speak(text)
    return {"message": "Text synthesized"}

@app.post("/listen")
async def listen():
    akulai.listen()
    return {"message": "Listening..."}
```
At that point I was able to start the program and then navigate to `http://localhost:8000/speak/Hello+There` to get it to say "Hello plus There". I guess it does not automatically urldecode the query string values.

Akul2010 commented:
I will add your proposed changes right now. I'm not sure how I would implement urldecode, however.

Akul2010 committed:
Fixed API set up not working

aaronchantrill commented:
I'm still having some issues with this. I'll try to submit a pull request. I'm trying to figure out why it won't shut down when I ask it to right now. Are you running this code on your system?

Akul2010 committed:
Automatically open the API link in browser

Akul2010 committed:
Merge branch 'Akul-AI:master-v2' into add-plugin-api-v2

Akul2010 committed:
Merge branch 'Akul-AI:master-v2' into add-plugin-api-v2

Akul2010 committed:
Merge branch 'Akul-AI:master-v2' into add-plugin-api-v2

Akul2010 committed:
Merge branch 'Akul-AI:master-v2' into add-plugin-api-v2

Akul2010 committed:
Replace `self` with `akulai` for objects outside of the `AkulAI` class

Refers to this code block:
```
# Create the listening thread
    akulai.stop_listening = threading.Event()
    akulai.listening_thread = threading.Thread(target=akulai.listen)
    akulai.listening_thread.start()
```
Since it is no longer in the `AkulAI` class, I made an object for it and used that instead.

Akul2010 committed:
Fix a line in the setup.bat script

We are very close to releasing 2.0, and so I wanted to double check everything, starting with scripts. However, I noticed so many things wrong with it. This is just fixing the setup.bat file, but I also want to cut things from other scripts. For example, I am considering putting plugins in here by default, rather than having to install them. (VOSK and all the other programs would still be in the script because they are too large to include here.)

Akul2010 committed:
Fix setup scripts

It seems the problem was that the `.gitmodules` file already had the submodule for plugins in it, so it wasn't working. It should now, although for updating plugins, I might switch from git submodules to something else.

Akul2010 committed:
Change LICENSE from MIT to GNU GPLv3

Akul2010 committed:
Update `README.md`

Akul2010 committed:
Removed tests folder because it was not needed

Akul2010 committed:
Add all default plugins to core repository and remove the plugin downloading from the scripts

Akul2010 committed:
Start fixinng plugins and getting server running

Akul2010 committed:
Update README.md

Akul2010 committed:
Update .gitignore

Akul2010 committed:
Ignore all vosk model files for testing

Akul2010 committed:
Merge branch 'master-v2' into add-plugin-api-v2

Akul2010 commented:
There seems to be a problem where the plugins can't access the FastAPI server. I'm not sure what is causing this, any ideas?

Akul2010 committed:
Shift API setup into the if condition at the end

Akul2010 committed:
Add plugin files

Akul2010 committed:
Fix VOSK model not being able to read the model files

Akul2010 committed:
Update `.idea` folder for anaconda distribution

Akul2010 committed:
Remove packages we don't use

Akul2010 committed:
Stop the `Error loading ASGI app. Attribute "app" not found in module "akulai". ` error

For some reason, until a keyboard interrupt, the VOSK model and FastAPI keep initializing themselves. I haven't been able to test after that, however.

Akul2010 committed:
Update akulai.py

Akul2010 committed:
update js plugins

Akul2010 committed:
Fixed all threading related errors, some other errors still remain

Akul2010 committed:
Strawberry Perl docs in short for new users

Akul2010 committed:
No more Pylance errors, but plugins don't work

Akul2010 commented:
Rather than calling through a server, I'm thinking we could just wrap the functions into a command line interface, through which the respective languages will have their own functions for. (for example, JS listen function will run a command in js to run the listen function)
